Sources/SwiftConcurrentSequence/Array+ConcurrentMap.swift

```diff
@@ -0,0 +1,70 @@
+//
+//  Created by Dan Federman on 4/30/22.
+//  Copyright © 2022 Dan Federman.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS"BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Dispatch
+
+extension Array {
+    /// Returns an array containing the results of mapping the given closure
+    /// over the sequence's elements. The given closure is executed concurrently
+    /// on multiple queues to reduce the wall-time consumed by the transform.
+    ///
+    /// - Parameter transform: A mapping closure. `transform` accepts an
+    ///   element of this sequence as its parameter and returns a transformed
+    ///   value of the same or of a different type.
+    /// - Returns: An array containing the transformed elements of this
+    ///   sequence.
+    public func concurrentMap<T>(_ transform: (Element) -> T) -> [T] {
+        // Create a buffer where we can store the transformed output.
+        var transformed = [T?](repeating: nil, count: count)
+        // Access the underlying array memory to concurrently write to indices.
+        return transformed.withUnsafeMutableBufferPointer { bufferPointer in
+            DispatchQueue.concurrentPerform(iterations: count) { index in
+                // It is safe to concurrently write to unique indexes of the buffer pointer.
+                bufferPointer[index] = transform(self[index])
+            }
+            return bufferPointer.compactMap { $0 }
+        }
+    }
+
+    /// Returns an array containing the results of mapping the given closure
+    /// over the sequence's elements. The given closure is executed concurrently
+    /// on multiple queues to reduce the wall-time consumed by the transform.
+    ///
+    /// - Parameter transform: A mapping closure. `transform` accepts an
+    ///   element of this sequence as its parameter and returns a transformed
+    ///   value of the same or of a different type.
+    /// - Returns: An array containing the transformed elements of this
+    ///   sequence.
+    public func concurrentMap<T>(_ transform: @escaping (Element) throws -> T) async rethrows -> [T] {
+        try await withThrowingTaskGroup(
```

Comment:
it'd be faster to 
```
await Task {
  concurrentMap(transform: transform)
}
```

However, we'd lose the ability to `throw` from the `transform` if we did that since we can't `throw` within `DispatchQueue.concurrentPerform`. I optimized for preserving language features in this `async` API over optimizing for speed. Curious to hear if that feels like the right optimization.

My thinking here is that structured concurrency optimizes for safety over speed, so optimizing for enabling error propagation over speed seemed reasonable within that framing.

Reply:
My guess is that the wall-time of synchronous execution will be faster than waiting for the scheduler to come back and enqueue each `await`ed array operations. But I could be wrong!

Reply:
Indeed! I really wish `measure` worked with `await`. In the interim I may have to build something to get reasonable side-by-side perf tests.